src/write_stage.rs

```diff
@@ -38,32 +44,97 @@ impl WriteStage {
         let entries = entry_receiver.recv_timeout(Duration::new(1, 0))?;
         let votes = entries_to_votes(&entries);
         crdt.write().unwrap().insert_votes(&votes);
+
+        //TODO(anatoly): real stake based voting needs to change this
+        //leader simply votes if the current set of validators have voted
+        //on a valid last id
         entry_writer.write_and_register_entries(&entries)?;
         trace!("New blobs? {}", entries.len());
         let mut blobs = VecDeque::new();
         entries.to_blobs(blob_recycler, &mut blobs);
+
         if !blobs.is_empty() {
-            inc_new_counter!("write_stage-broadcast_vote-count", votes.len());
-            inc_new_counter!("write_stage-broadcast_blobs-count", blobs.len());
+            inc_new_counter!("write_stage-recv_vote", votes.len());
+            inc_new_counter!("write_stage-broadcast_blobs", blobs.len());
             trace!("broadcasting {}", blobs.len());
             blob_sender.send(blobs)?;
         }
         Ok(())
     }
-
+    pub fn leader_vote(
+        debug_id: u64,
+        keypair: &KeyPair,
+        bank: &Arc<Bank>,
+        crdt: &Arc<RwLock<Crdt>>,
+        blob_recycler: &BlobRecycler,
+        vote_blob_sender: &BlobSender,
+        last_vote: &mut u64,
+    ) -> Result<()> {
+        let now = timing::timestamp();
+        if now - *last_vote > VOTE_TIMEOUT_MS {
+            //TODO(anatoly): vote if the last id set is mostly valid
+            let ids: Vec<_> = crdt.read()
+                .unwrap()
+                .table
+                .values()
+                .map(|x| x.ledger_state.last_id)
+                .collect();
+            let total = bank.count_valid_ids(&ids);
+            //TODO(anatoly): this isn't stake based voting
+            info!(
+                "{:x}: valid_ids {}/{} {}",
+                debug_id,
+                total,
+                ids.len(),
+                (2 * ids.len()) / 3
+            );
+            if total > (2 * ids.len()) / 3 {
```

Comment:
Should there be a test showing that this must be `>` and not `>=`? At a glance, it seems like it should be `>=` and that the leader is the one to make it `2/3 + 1`.